# Bodyfat
Το 2ο dataset που τρέχω SVR είναι το bodyfat (https://lib.stat.cmu.edu/datasets/bodyfat), στο οποίο προσπαθούμε να υπολογίσουμε το ποσοστό λίπους σε ανθρώπους δοσμένων σωματικών τους καρακτηριστικών, όπως βάρος, ύψος και μήκος περιφέρειας διαφόρων οργάνων (πχ. λαιμού ή κοιλιάς). Το dataset το κατέβασα από το παραπάνω link και εφαρμόζω τεχνικές RegEx για να κάνω extract τα δεδομένα:

In [38]:
import numpy as np
import pandas as pd

In [81]:
with open('bodyfats.txt', 'r') as f:
    bodyfats_str = f.read()   
#print(bodyfats_str)


In [27]:
import re

# match all numerical values and discard whitespaces and \n chars
numbers_str_list = re.findall(r'\d+\.\d+|\d+', bodyfats_str)

# Convert the list of strings to a list of integers
flattened_dset = np.array([float(num) for num in numbers_str_list])

print(flattened_dset)

[ 1.0708 12.3    23.     ... 33.7    30.     20.9   ]


In [28]:
'''
The variables listed below, from left to right, are:

 Density determined from underwater weighing
 Percent body fat from Siri's (1956) equation
 Age (years)
 Weight (lbs)
 Height (inches)
 Neck circumference (cm)
 Chest circumference (cm)
 Abdomen 2 circumference (cm)
 Hip circumference (cm)
 Thigh circumference (cm)
 Knee circumference (cm)
 Ankle circumference (cm)
 Biceps (extended) circumference (cm)
 Forearm circumference (cm)
 Wrist circumference (cm)
'''

"\nThe variables listed below, from left to right, are:\n\n Density determined from underwater weighing\n Percent body fat from Siri's (1956) equation\n Age (years)\n Weight (lbs)\n Height (inches)\n Neck circumference (cm)\n Chest circumference (cm)\n Abdomen 2 circumference (cm)\n Hip circumference (cm)\n Thigh circumference (cm)\n Knee circumference (cm)\n Ankle circumference (cm)\n Biceps (extended) circumference (cm)\n Forearm circumference (cm)\n Wrist circumference (cm)\n"

In [29]:
num_features = 15
num_samples = len(flattened_dset) // num_features

print(num_features, num_samples)

15 252


Έχουμε 13 ανεξάρτητες μεταβλητές και 1 εξαρτημένη (την Percent body fat), με συνολικά 252 δείγματα.

In [42]:
df_arr = flattened_dset.reshape(num_samples, num_features)

df = pd.DataFrame(df_arr, columns=['Density', 'Fat %', 'Age', 'Weight', 'Height',
                'Neck', 'Chest', 'Abdomen', 'Hip', 'Knee', 'Thigh', 'Ankle', 'Bicepts', 'Forearm', 'Wrist'])

In [43]:
df.head()

Density  Fat %   Age  Weight  Height  Neck  Chest  Abdomen    Hip  Knee  \
0   1.0708   12.3  23.0  154.25   67.75  36.2   93.1     85.2   94.5  59.0   
1   1.0853    6.1  22.0  173.25   72.25  38.5   93.6     83.0   98.7  58.7   
2   1.0414   25.3  22.0  154.00   66.25  34.0   95.8     87.9   99.2  59.6   
3   1.0751   10.4  26.0  184.75   72.25  37.4  101.8     86.4  101.2  60.1   
4   1.0340   28.7  24.0  184.25   71.25  34.4   97.3    100.0  101.9  63.2   

   Thigh  Ankle  Bicepts  Forearm  Wrist  
0   37.3   21.9     32.0     27.4   17.1  
1   37.3   23.4     30.5     28.9   18.2  
2   38.9   24.0     28.8     25.2   16.6  
3   37.3   22.8     32.4     29.4   18.2  
4   42.2   24.0     32.2     27.7   17.7

In [44]:
df.describe()

Density       Fat %         Age      Weight      Height  \
count   252.000000  252.000000  252.000000  252.000000  252.000000   
mean     40.535752   19.175794   44.884921  178.924405   70.148810   
std     626.724575    8.322686   12.602040   29.389160    3.662856   
min       1.010100    0.000000   22.000000  118.500000   29.500000   
25%       1.041550   12.475000   35.750000  159.000000   68.250000   
50%       1.054950   19.200000   43.000000  176.500000   70.000000   
75%       1.070425   25.300000   54.000000  197.000000   72.250000   
max    9950.000000   47.500000   81.000000  363.150000   77.750000   

             Neck       Chest     Abdomen         Hip        Knee       Thigh  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean    37.992063  100.824206   92.555952   99.904762   59.405952   38.590476   
std      2.430913    8.430476   10.783077    7.164058    5.249952    2.411805   
min     31.100000   79.300000   69.400000   85.000000   47.200000   33.000000   
25%     36.400000   94.350000   84.575000   95.500000   56.000000   36.975000   
50%     38.000000   99.650000   90.950000   99.300000   59.000000   38.500000   
75%     39.425000  105.375000   99.325000  103.525000   62.350000   39.925000   
max     51.200000  136.200000  148.100000  147.700000   87.300000   49.100000   

            Ankle     Bicepts     Forearm       Wrist  
count  252.000000  252.000000  252.000000  252.000000  
mean    23.102381   32.273413   28.663889   18.229762  
std      1.694893    3.021274    2.020691    0.933585  
min     19.100000   24.800000   21.000000   15.800000  
25%     22.000000   30.200000   27.300000   17.600000  
50%     22.800000   32.050000   28.700000   18.300000  
75%     24.000000   34.325000   30.000000   18.800000  
max     33.900000   45.000000   34.900000   21.400000

Έπειτα από ανάλυση, παρατηρούμε ότι σε γενικές γραμμές τα δεδομένα είναι καθαρά. Το μόνο πρόβλημα που έχουμε είναι ένας outlier στην τιμή Density, τον οποίο αντικαθιστώ με τον μέσο όρο των samples.

In [59]:
df.iloc[215, 0] = 1.044950 # replace the outlier with mean

In [60]:
df.describe()

Density       Fat %         Age      Weight      Height        Neck  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean     1.055772   19.175794   44.884921  178.924405   70.148810   37.992063   
std      0.018654    8.322686   12.602040   29.389160    3.662856    2.430913   
min      1.010100    0.000000   22.000000  118.500000   29.500000   31.100000   
25%      1.041550   12.475000   35.750000  159.000000   68.250000   36.400000   
50%      1.054900   19.200000   43.000000  176.500000   70.000000   38.000000   
75%      1.070400   25.300000   54.000000  197.000000   72.250000   39.425000   
max      1.108900   47.500000   81.000000  363.150000   77.750000   51.200000   

            Chest     Abdomen         Hip        Knee       Thigh       Ankle  \
count  252.000000  252.000000  252.000000  252.000000  252.000000  252.000000   
mean   100.824206   92.555952   99.904762   59.405952   38.590476   23.102381   
std      8.430476   10.783077    7.164058    5.249952    2.411805    1.694893   
min     79.300000   69.400000   85.000000   47.200000   33.000000   19.100000   
25%     94.350000   84.575000   95.500000   56.000000   36.975000   22.000000   
50%     99.650000   90.950000   99.300000   59.000000   38.500000   22.800000   
75%    105.375000   99.325000  103.525000   62.350000   39.925000   24.000000   
max    136.200000  148.100000  147.700000   87.300000   49.100000   33.900000   

          Bicepts     Forearm       Wrist  
count  252.000000  252.000000  252.000000  
mean    32.273413   28.663889   18.229762  
std      3.021274    2.020691    0.933585  
min     24.800000   21.000000   15.800000  
25%     30.200000   27.300000   17.600000  
50%     32.050000   28.700000   18.300000  
75%     34.325000   30.000000   18.800000  
max     45.000000   34.900000   21.400000

In [62]:
X, y = df.drop(['Density', 'Fat %'], axis=1), df['Fat %']

In [65]:
y

0      12.3
1       6.1
2      25.3
3      10.4
4      28.7
       ... 
247    11.0
248    33.6
249    29.3
250    26.0
251    31.9
Name: Fat %, Length: 252, dtype: float64

Για να βοηθήσω το μοντέλο να τρέξει καλύτερα, κάνω MinMax scaling σε όλες τις στήλες, έτσι ώστε να έχουμε τιμές 0 ως 1, όπου κάνουμε map το min σε 0 και το max σε 1 με γραμμική αναλογία με την χρήση του εξής τύπου:
$$ MinMax(d_i) = \frac{d_i - min(d)}{max(d) - min(d)} $$

In [76]:
from sklearn.preprocessing import MinMaxScaler

X_norm = MinMaxScaler().fit_transform(X)

Χωρίζω το dataset 60-40 σε train-test και το train 80-20 σε train-validation όπως και στο άλλο dataset:

In [77]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_new, y, test_size=0.4, random_state=0) # 60-40 split

X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=0) # 80-20 split

Χρησιμοποιώ Linear Regression ως benchmark:

In [78]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_val)

print(f'LR MAE: {mean_absolute_error(y_val, y_pred)}')

LR MAE: 3.2189147287337505


### Linear Kernel
Ξεκινάω με την εφαρμογή ενός linear kernel ψάχνοντας καλές τιμές για τα c και ε. Θα κάνω διαδοχικά grid searches, κάθε φορά μεγαλώνοντας το granularity των παραμέτρων με βάση τις καλύτερες τιμές από την προηγούμενη επανάληψη.

In [79]:
from sklearn.svm import SVR

In [80]:
for C in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
    for epsilon in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
        model = SVR(kernel='linear', C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=0.001, e=0.001: 6.272035515027121
MAE C=0.001, e=0.01: 6.272035515027121
MAE C=0.001, e=0.1: 6.27214691064116
MAE C=0.001, e=1: 6.311124917129755
MAE C=0.001, e=10: 6.259725569362329
MAE C=0.001, e=100: 8.403225806451614
MAE C=0.001, e=1000: 8.403225806451628
MAE C=0.01, e=0.001: 6.24032499803582
MAE C=0.01, e=0.01: 6.24032499803582
MAE C=0.01, e=0.1: 6.239211041895446
MAE C=0.01, e=1: 6.28060401000723
MAE C=0.01, e=10: 6.245642790397483
MAE C=0.01, e=100: 8.403225806451614
MAE C=0.01, e=1000: 8.403225806451628
MAE C=0.1, e=0.001: 5.9772347315817544
MAE C=0.1, e=0.01: 5.9772347315817544
MAE C=0.1, e=0.1: 5.9772347315817544
MAE C=0.1, e=1: 5.993073009522078
MAE C=0.1, e=10: 6.085031297822443
MAE C=0.1, e=100: 8.403225806451614
MAE C=0.1, e=1000: 8.403225806451628
MAE C=1, e=0.001: 5.234521225421599
MAE C=1, e=0.01: 5.235095258359531
MAE C=1, e=0.1: 5.233512586072449
MAE C=1, e=1: 5.1887601289130245
MAE C=1, e=10: 5.598935849755717
MAE C=1, e=100: 8.403225806451614
MAE C=1, e=1000:

In [82]:
# MAE C=100, e=1: 3.280676857654416
for C in [80, 90, 100, 110, 120]:
    for epsilon in [.8, .9, 1, 1.1, 1.2]:
        model = SVR(kernel='linear', C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=80, e=0.8: 3.33509582163575
MAE C=80, e=0.9: 3.305031179155931
MAE C=80, e=1: 3.2943983804082064
MAE C=80, e=1.1: 3.2781567808686467
MAE C=80, e=1.2: 3.2337784778365144
MAE C=90, e=0.8: 3.3292588247353416
MAE C=90, e=0.9: 3.2985637794050273
MAE C=90, e=1: 3.284941511504564
MAE C=90, e=1.1: 3.281106651126427
MAE C=90, e=1.2: 3.2389184418664
MAE C=100, e=0.8: 3.313807802553015
MAE C=100, e=0.9: 3.2848869642395755
MAE C=100, e=1: 3.280676857654416
MAE C=100, e=1.1: 3.27872226155978
MAE C=100, e=1.2: 3.2299461456892313
MAE C=110, e=0.8: 3.308991442656644
MAE C=110, e=0.9: 3.287125983243364
MAE C=110, e=1: 3.2810118534406096
MAE C=110, e=1.1: 3.2856428441521803
MAE C=110, e=1.2: 3.2387802162071213
MAE C=120, e=0.8: 3.3087502135893434
MAE C=120, e=0.9: 3.293888703798792
MAE C=120, e=1: 3.2916121905176086
MAE C=120, e=1.1: 3.3023946879253754
MAE C=120, e=1.2: 3.2513767834636194


In [83]:
for C in [50, 60, 70, 75]:
    for epsilon in [1.2, 1.3, 1.4, 1.5]:
        model = SVR(kernel='linear', C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=50, e=1.2: 3.285812694520106
MAE C=50, e=1.3: 3.274209737796452
MAE C=50, e=1.4: 3.231532882309553
MAE C=50, e=1.5: 3.214827600054024
MAE C=60, e=1.2: 3.2213824838132106
MAE C=60, e=1.3: 3.2008287495942622
MAE C=60, e=1.4: 3.2001238771977625
MAE C=60, e=1.5: 3.1997239622942626
MAE C=70, e=1.2: 3.2311531892685865
MAE C=70, e=1.3: 3.2075432200703378
MAE C=70, e=1.4: 3.161422860363879
MAE C=70, e=1.5: 3.1703146702939753
MAE C=75, e=1.2: 3.227340222943265
MAE C=75, e=1.3: 3.2064811728743194
MAE C=75, e=1.4: 3.1476222988783618
MAE C=75, e=1.5: 3.1536361091752076


Για το linear kernel, η καλύτερη τιμή που παίρνουμε είναι ΜΑΕ περίπου 3.1476 για παραμέτρους `C=75, e=1.4`.

### Πολυωνυμικά kernel
Συνεχίζω με την χρήση πολυωνυμικών kernels βαθμόυ n=2,3,5,7,9 με παρόμοιο τρόπο όπως με τα linear kernels.

#### 2nd order

In [84]:
for C in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
    for epsilon in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
        model = SVR(kernel='poly', degree=2, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=0.001, e=0.001: 6.201926899435834
MAE C=0.001, e=0.01: 6.201926899435834
MAE C=0.001, e=0.1: 6.201926899435834
MAE C=0.001, e=1: 6.244077779592154
MAE C=0.001, e=10: 6.227761825233908
MAE C=0.001, e=100: 8.403225806451614
MAE C=0.001, e=1000: 8.403225806451628
MAE C=0.01, e=0.001: 5.8723946579439845
MAE C=0.01, e=0.01: 5.8723946579439845
MAE C=0.01, e=0.1: 5.8723946579439845
MAE C=0.01, e=1: 5.900418612874061
MAE C=0.01, e=10: 5.9964086395669485
MAE C=0.01, e=100: 8.403225806451614
MAE C=0.01, e=1000: 8.403225806451628
MAE C=0.1, e=0.001: 5.131390550073381
MAE C=0.1, e=0.01: 5.132349496594657
MAE C=0.1, e=0.1: 5.144080571114534
MAE C=0.1, e=1: 5.1336793686372895
MAE C=0.1, e=10: 5.639423825171895
MAE C=0.1, e=100: 8.403225806451614
MAE C=0.1, e=1000: 8.403225806451628
MAE C=1, e=0.001: 4.110540115734542
MAE C=1, e=0.01: 4.1074846924987405
MAE C=1, e=0.1: 4.0842731240108705
MAE C=1, e=1: 3.9694446996876898
MAE C=1, e=10: 4.973118241093455
MAE C=1, e=100: 8.403225806451614
MAE C=1,

In [85]:
# MAE C=10, e=0.1: 3.5732334108724917
for C in [8, 9, 10, 11, 12]:
    for epsilon in [.08, .09, .1, .11, .12]:
        model = SVR(kernel='poly', degree=2, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=8, e=0.08: 3.51279536339849
MAE C=8, e=0.09: 3.511207502286187
MAE C=8, e=0.1: 3.50957353631428
MAE C=8, e=0.11: 3.508009830374045
MAE C=8, e=0.12: 3.506386571562228
MAE C=9, e=0.08: 3.548011600594835
MAE C=9, e=0.09: 3.546098434618395
MAE C=9, e=0.1: 3.5441602381422337
MAE C=9, e=0.11: 3.5422687364083547
MAE C=9, e=0.12: 3.540520212042529
MAE C=10, e=0.08: 3.5777049961087664
MAE C=10, e=0.09: 3.575345658716354
MAE C=10, e=0.1: 3.5732334108724917
MAE C=10, e=0.11: 3.5712907147888915
MAE C=10, e=0.12: 3.569568106310117
MAE C=11, e=0.08: 3.5963045850345106
MAE C=11, e=0.09: 3.5934430148052305
MAE C=11, e=0.1: 3.5910650601915957
MAE C=11, e=0.11: 3.58946094207945
MAE C=11, e=0.12: 3.587767425440526
MAE C=12, e=0.08: 3.614408630854416
MAE C=12, e=0.09: 3.612013066360422
MAE C=12, e=0.1: 3.6108574526196793
MAE C=12, e=0.11: 3.6096987468908646
MAE C=12, e=0.12: 3.6085260275626387


In [88]:
# MAE C=8, e=0.12: 3.506386571562228
for C in [4, 5, 6, 7, 7.5]:
    for epsilon in [.14, .15, .16, .17, .18]:
        model = SVR(kernel='poly', degree=2, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=4, e=0.14: 3.4569769170309397
MAE C=4, e=0.15: 3.4596469727054404
MAE C=4, e=0.16: 3.4591376940950984
MAE C=4, e=0.17: 3.458691582219563
MAE C=4, e=0.18: 3.4587091748418977
MAE C=5, e=0.14: 3.449407784620497
MAE C=5, e=0.15: 3.4488556716195484
MAE C=5, e=0.16: 3.4483552581268695
MAE C=5, e=0.17: 3.447900507571375
MAE C=5, e=0.18: 3.4478939783159643
MAE C=6, e=0.14: 3.4385510805057122
MAE C=6, e=0.15: 3.4371916777768536
MAE C=6, e=0.16: 3.436908222825276
MAE C=6, e=0.17: 3.4366304909389847
MAE C=6, e=0.18: 3.4363318010144774
MAE C=7, e=0.14: 3.473351642523406
MAE C=7, e=0.15: 3.4702410899799925
MAE C=7, e=0.16: 3.4672236282840587
MAE C=7, e=0.17: 3.4642296800965275
MAE C=7, e=0.18: 3.461258942078079
MAE C=7.5, e=0.14: 3.4838256933081273
MAE C=7.5, e=0.15: 3.480825339822489
MAE C=7.5, e=0.16: 3.4778630096337304
MAE C=7.5, e=0.17: 3.4751841989816716
MAE C=7.5, e=0.18: 3.4730797757498437


Για poly kernel 2ου βαθμόυ καλύτερη απόδοση παίρνουμε για `C=7, e=0.18` με ΜΑΕ περίπου 3.4613.

#### 3rd order

In [89]:
for C in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
    for epsilon in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
        model = SVR(kernel='poly', degree=3, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=0.001, e=0.001: 5.96821589367854
MAE C=0.001, e=0.01: 5.967227300161483
MAE C=0.001, e=0.1: 5.95734136499092
MAE C=0.001, e=1: 6.016337098218354
MAE C=0.001, e=10: 6.057572276836313
MAE C=0.001, e=100: 8.403225806451614
MAE C=0.001, e=1000: 8.403225806451628
MAE C=0.01, e=0.001: 5.469880862662359
MAE C=0.01, e=0.01: 5.470168821841505
MAE C=0.01, e=0.1: 5.471102719672352
MAE C=0.01, e=1: 5.441116647502514
MAE C=0.01, e=10: 5.7928732293476495
MAE C=0.01, e=100: 8.403225806451614
MAE C=0.01, e=1000: 8.403225806451628
MAE C=0.1, e=0.001: 4.415459455934141
MAE C=0.1, e=0.01: 4.417103655242877
MAE C=0.1, e=0.1: 4.42045699079714
MAE C=0.1, e=1: 4.300625036638203
MAE C=0.1, e=10: 5.251192159343065
MAE C=0.1, e=100: 8.403225806451614
MAE C=0.1, e=1000: 8.403225806451628
MAE C=1, e=0.001: 3.7320918150331805
MAE C=1, e=0.01: 3.7279854266592958
MAE C=1, e=0.1: 3.714079010817701
MAE C=1, e=1: 3.616839537460453
MAE C=1, e=10: 4.090495495679103
MAE C=1, e=100: 8.403225806451614
MAE C=1, e=1000:

In [91]:
# MAE C=1, e=1: 3.616839537460453
for C in [.8, .9, 1, 1.1, 1.2]:
    for epsilon in [0.8, 0.9, 1, 1.1, 1.2]:
        model = SVR(kernel='poly', degree=3, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=0.8, e=0.8: 3.7013443552889562
MAE C=0.8, e=0.9: 3.689672260346139
MAE C=0.8, e=1: 3.6851108284523613
MAE C=0.8, e=1.1: 3.6826640998930005
MAE C=0.8, e=1.2: 3.680282200352879
MAE C=0.9, e=0.8: 3.629801914454611
MAE C=0.9, e=0.9: 3.6116790699247816
MAE C=0.9, e=1: 3.619242785204236
MAE C=0.9, e=1.1: 3.630078273542888
MAE C=0.9, e=1.2: 3.647145550491861
MAE C=1, e=0.8: 3.617307038049931
MAE C=1, e=0.9: 3.6206994119765183
MAE C=1, e=1: 3.616839537460453
MAE C=1, e=1.1: 3.6313603341901666
MAE C=1, e=1.2: 3.641003911928963
MAE C=1.1, e=0.8: 3.696652718212665
MAE C=1.1, e=0.9: 3.661970065365002
MAE C=1.1, e=1: 3.6730101915759197
MAE C=1.1, e=1.1: 3.677705825791492
MAE C=1.1, e=1.2: 3.6500911614988114
MAE C=1.2, e=0.8: 3.7337286617348955
MAE C=1.2, e=0.9: 3.7275614029441666
MAE C=1.2, e=1: 3.7186879004278652
MAE C=1.2, e=1.1: 3.6938894436377674
MAE C=1.2, e=1.2: 3.657928772125294


Σε πολυωνυμικό kernel 3ου βαθμόυ έχουμε το μοντέλο `C=0.9, e=0.9` με ΜΑΕ 3.6117.
#### 5ου βαθμού

In [93]:
for C in [0.001, 0.01, 0.1, 1, 10, 100]:
    for epsilon in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
        model = SVR(kernel='poly', degree=5, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=0.001, e=0.001: 5.062351503791367
MAE C=0.001, e=0.01: 5.061130551425403
MAE C=0.001, e=0.1: 5.052137602077159
MAE C=0.001, e=1: 5.013904473843948
MAE C=0.001, e=10: 5.748174080161789
MAE C=0.001, e=100: 8.403225806451614
MAE C=0.001, e=1000: 8.403225806451628
MAE C=0.01, e=0.001: 4.311760775362377
MAE C=0.01, e=0.01: 4.312588550403395
MAE C=0.01, e=0.1: 4.322300976165233
MAE C=0.01, e=1: 4.283622323540498
MAE C=0.01, e=10: 4.752107014751086
MAE C=0.01, e=100: 8.403225806451614
MAE C=0.01, e=1000: 8.403225806451628
MAE C=0.1, e=0.001: 3.9438938837975885
MAE C=0.1, e=0.01: 3.9400809437307314
MAE C=0.1, e=0.1: 3.917857562260511
MAE C=0.1, e=1: 3.869126342976727
MAE C=0.1, e=10: 4.314442540888004
MAE C=0.1, e=100: 8.403225806451614
MAE C=0.1, e=1000: 8.403225806451628
MAE C=1, e=0.001: 4.33896447610124
MAE C=1, e=0.01: 4.335630387838681
MAE C=1, e=0.1: 4.299948186991823
MAE C=1, e=1: 3.9931317113272424
MAE C=1, e=10: 4.201778357748185
MAE C=1, e=100: 8.403225806451614
MAE C=1, e=100

In [94]:
# MAE C=0.1, e=1: 3.869126342976727
for C in [.08, .09, 0.1, .11, .12]:
    for epsilon in [0.8, 0.9, 1, 1.1, 1.2]:
        model = SVR(kernel='poly', degree=5, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=0.08, e=0.8: 3.956315015039417
MAE C=0.08, e=0.9: 3.9444220078664203
MAE C=0.08, e=1: 3.927423724785232
MAE C=0.08, e=1.1: 3.897232114918752
MAE C=0.08, e=1.2: 3.863316165304378
MAE C=0.09, e=0.8: 3.93696508289676
MAE C=0.09, e=0.9: 3.9171596023578323
MAE C=0.09, e=1: 3.896345721688826
MAE C=0.09, e=1.1: 3.875706998694865
MAE C=0.09, e=1.2: 3.843998954737907
MAE C=0.1, e=0.8: 3.9044434622174977
MAE C=0.1, e=0.9: 3.8894833621626113
MAE C=0.1, e=1: 3.869126342976727
MAE C=0.1, e=1.1: 3.8493253046765292
MAE C=0.1, e=1.2: 3.8231928262732935
MAE C=0.11, e=0.8: 3.889814348080069
MAE C=0.11, e=0.9: 3.863863384778667
MAE C=0.11, e=1: 3.8496631068687495
MAE C=0.11, e=1.1: 3.8357765888468176
MAE C=0.11, e=1.2: 3.8129176342937847
MAE C=0.12, e=0.8: 3.8782734469778837
MAE C=0.12, e=0.9: 3.849212614201847
MAE C=0.12, e=1: 3.8280273415639865
MAE C=0.12, e=1.1: 3.8242846876250374
MAE C=0.12, e=1.2: 3.818667730021499


Σε πολυωνυμικό kernel 5ου βαθμόυ έχουμε το μοντέλο `C=0.11, e=1.2` με ΜΑΕ 3.813.

#### Μεγαλύτερου βαθμού

In [95]:
# degree 7
for C in [0.001, 0.01, 0.1, 1, 10]: # avoid larger values, no better results and longer execution times
    for epsilon in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
        model = SVR(kernel='poly', degree=7, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=0.001, e=0.001: 4.314859180207518
MAE C=0.001, e=0.01: 4.314190798871022
MAE C=0.001, e=0.1: 4.306888761703554
MAE C=0.001, e=1: 4.235476983634937
MAE C=0.001, e=10: 5.006919497617695
MAE C=0.001, e=100: 8.403225806451614
MAE C=0.001, e=1000: 8.403225806451628
MAE C=0.01, e=0.001: 4.209960650224699
MAE C=0.01, e=0.01: 4.209342307752513
MAE C=0.01, e=0.1: 4.203060838016605
MAE C=0.01, e=1: 4.262358789982092
MAE C=0.01, e=10: 4.602177644819971
MAE C=0.01, e=100: 8.403225806451614
MAE C=0.01, e=1000: 8.403225806451628
MAE C=0.1, e=0.001: 5.022594525261896
MAE C=0.1, e=0.01: 5.020524401354345
MAE C=0.1, e=0.1: 4.991966290368661
MAE C=0.1, e=1: 4.6879799443769885
MAE C=0.1, e=10: 4.949963049935688
MAE C=0.1, e=100: 8.403225806451614
MAE C=0.1, e=1000: 8.403225806451628
MAE C=1, e=0.001: 9.906790541511443
MAE C=1, e=0.01: 9.896862183881307
MAE C=1, e=0.1: 9.795894028476372
MAE C=1, e=1: 8.541678422573133
MAE C=1, e=10: 4.949963049935688
MAE C=1, e=100: 8.403225806451614
MAE C=1, e=1000

In [101]:
# degree 9
for C in [1e-4, 0.001, 0.01, 0.1]: # avoid larger values, no better results and longer execution times
    for epsilon in [1e-4, 0.001, 0.01, 0.1, 1, 10, 100, 1000]:
        model = SVR(kernel='poly', degree=9, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=0.0001, e=0.0001: 5.098825066652949
MAE C=0.0001, e=0.001: 5.098579339010363
MAE C=0.0001, e=0.01: 5.096418458107743
MAE C=0.0001, e=0.1: 5.08005840115008
MAE C=0.0001, e=1: 4.9934878013553226
MAE C=0.0001, e=10: 5.2953831862875145
MAE C=0.0001, e=100: 8.403225806451614
MAE C=0.0001, e=1000: 8.403225806451628
MAE C=0.001, e=0.0001: 4.4536325876636145
MAE C=0.001, e=0.001: 4.45386872891053
MAE C=0.001, e=0.01: 4.456971387024118
MAE C=0.001, e=0.1: 4.48410495770439
MAE C=0.001, e=1: 4.576219551520446
MAE C=0.001, e=10: 5.689782251823135
MAE C=0.001, e=100: 8.403225806451614
MAE C=0.001, e=1000: 8.403225806451628
MAE C=0.01, e=0.0001: 11.7491093484438
MAE C=0.01, e=0.001: 11.748383497794926
MAE C=0.01, e=0.01: 11.741012642051377
MAE C=0.01, e=0.1: 11.658719003568487
MAE C=0.01, e=1: 10.75618517989808
MAE C=0.01, e=10: 6.101652947272385
MAE C=0.01, e=100: 8.403225806451614
MAE C=0.01, e=1000: 8.403225806451628
MAE C=0.1, e=0.0001: 12.028166562367476
MAE C=0.1, e=0.001: 12.02485096479

In [97]:
# degree 13
for C in [0.001, 0.01, 0.1]: # avoid larger values, no better results and longer execution times
    for epsilon in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
        model = SVR(kernel='poly', degree=13, C=C, epsilon=epsilon)
        model.fit(X_train, y_train)

        y_pred = model.predict(X_val)

        print(f'MAE C={C}, e={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE C=0.001, e=0.001: 40.91402978578563
MAE C=0.001, e=0.01: 40.955922933576616
MAE C=0.001, e=0.1: 41.36649513478771
MAE C=0.001, e=1: 46.457421226898965
MAE C=0.001, e=10: 14.416531277847998
MAE C=0.001, e=100: 8.403225806451614
MAE C=0.001, e=1000: 8.403225806451628
MAE C=0.01, e=0.001: 137.82654230013384
MAE C=0.01, e=0.01: 137.7926664934905
MAE C=0.01, e=0.1: 137.62264226675873
MAE C=0.01, e=1: 129.5479931574922
MAE C=0.01, e=10: 14.416531277847998
MAE C=0.01, e=100: 8.403225806451614
MAE C=0.01, e=1000: 8.403225806451628
MAE C=0.1, e=0.001: 151.32221092014535
MAE C=0.1, e=0.01: 151.1808660791023
MAE C=0.1, e=0.1: 149.75913844726833
MAE C=0.1, e=1: 136.04034052198503
MAE C=0.1, e=10: 14.416531277847998
MAE C=0.1, e=100: 8.403225806451614
MAE C=0.1, e=1000: 8.403225806451628


Γενικά πιάνουμε άσχημες αποδόσεις για πολυωνυμικά kernels μεγάλου βαθμού. Πριν προχωρήσω στα RBF, δοκιμάζω τα καλύτερα μοντέλα που πήρα από το validation loss στο test set.

In [103]:
def train_and_test(model):
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    
    return mean_absolute_error(y_test, y_pred)

# Linear Regression
best_lr = LinearRegression()
print(f'Linear Regression MAE: {train_and_test(best_linear)}')

# Linear
best_linear = SVR(kernel='linear', C=75, epsilon=1.4)
print(f'Linear C=75, e=1.4 MAE: {train_and_test(best_linear)}')

# 2nd order
best_poly2 = SVR(kernel='poly', degree=2, C=7, epsilon=0.18)
print(f'Best SVR poly deg 2 MAE: {train_and_test(best_poly2)}')

# 3rd order
best_poly3 = SVR(kernel='poly', degree=3, C=0.9, epsilon=0.9)
print(f'Best SVR poly deg 3 MAE: {train_and_test(best_poly3)}')

# 5th order
best_poly5 = SVR(kernel='poly', degree=5, C=.11, epsilon=1.2)
print(f'Best SVR poly deg 5 MAE: {train_and_test(best_poly5)}')

# 7th order
best_poly7 = SVR(kernel='poly', degree=7, C=.01, epsilon=0.1)
print(f'Best SVR poly deg 7 MAE: {train_and_test(best_poly7)}')

# 9th order
best_poly9 = SVR(kernel='poly', degree=9, C=.001, epsilon=0.0001)
print(f'Best SVR poly deg 9 MAE: {train_and_test(best_poly9)}')

Linear Regression MAE: 3.629658592421555
Linear C=75, e=1.4 MAE: 3.629658592421555
Best SVR poly deg 2 MAE: 3.847261719645741
Best SVR poly deg 3 MAE: 3.892270391345038
Best SVR poly deg 5 MAE: 4.828815058229227
Best SVR poly deg 7 MAE: 6.216640039878661
Best SVR poly deg 9 MAE: 10.689284311104888


Παρατηρούμε ότι η το μοντέλο της γραμμικής παλινδρόμησης έχει ακριβώς την ίδια απόδοση με το Linear SVM στο test set και ότι όσο ανεβάζουμε τον βαθμό πολυωνύμου στο kernel, τόσο χειροτερεύει η απόδοση.

### RBF kernel
Ακολουθούμε παρόμοια στρατιγική fine tuning με τα polynomial kernel μοντέλα, μόνο που εδώ θα πρέπει να έχουμε 3D grid search προσθέτοντας το gamma.

In [104]:
for gamma in [100, 10, 1, 0.1, 0.01, 0.001]:
    for C in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
        for epsilon in [0.001, 0.01, 0.1, 1, 10, 100, 1000]:
            model = SVR(kernel='rbf', C=C, epsilon=epsilon, gamma=gamma)
            model.fit(X_train, y_train)

            y_pred = model.predict(X_val)

            print(f'MAE γ={gamma}, C={C}, ε={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE γ=100, C=0.001, ε=0.001: 6.275803180057625
MAE γ=100, C=0.001, ε=0.01: 6.275803180057625
MAE γ=100, C=0.001, ε=0.1: 6.275802914002435
MAE γ=100, C=0.001, ε=1: 6.314547919492007
MAE γ=100, C=0.001, ε=10: 6.261291279273535
MAE γ=100, C=0.001, ε=100: 8.403225806451614
MAE γ=100, C=0.001, ε=1000: 8.403225806451628
MAE γ=100, C=0.01, ε=0.001: 6.275773736060126
MAE γ=100, C=0.01, ε=0.01: 6.275773736060126
MAE γ=100, C=0.01, ε=0.1: 6.275771075508239
MAE γ=100, C=0.01, ε=1: 6.314834033629743
MAE γ=100, C=0.01, ε=10: 6.261299889509548
MAE γ=100, C=0.01, ε=100: 8.403225806451614
MAE γ=100, C=0.01, ε=1000: 8.403225806451628
MAE γ=100, C=0.1, ε=0.001: 6.275464887447046
MAE γ=100, C=0.1, ε=0.01: 6.275462492950348
MAE γ=100, C=0.1, ε=0.1: 6.275452690566277
MAE γ=100, C=0.1, ε=1: 6.3175586567505695
MAE γ=100, C=0.1, ε=10: 6.2613859918696795
MAE γ=100, C=0.1, ε=100: 8.403225806451614
MAE γ=100, C=0.1, ε=1000: 8.403225806451628
MAE γ=100, C=1, ε=0.001: 6.2857157696293084
MAE γ=100, C=1, ε=0.01: 6.2

MAE γ=0.01, C=0.001, ε=0.001: 6.275731829123107
MAE γ=0.01, C=0.001, ε=0.01: 6.275731829123107
MAE γ=0.01, C=0.001, ε=0.1: 6.275734035152294
MAE γ=0.01, C=0.001, ε=1: 6.314448945715264
MAE γ=0.01, C=0.001, ε=10: 6.261259426918075
MAE γ=0.01, C=0.001, ε=100: 8.403225806451614
MAE γ=0.01, C=0.001, ε=1000: 8.403225806451628
MAE γ=0.01, C=0.01, ε=0.001: 6.275060226714933
MAE γ=0.01, C=0.01, ε=0.01: 6.275060226714933
MAE γ=0.01, C=0.01, ε=0.1: 6.275082287006823
MAE γ=0.01, C=0.01, ε=1: 6.3138442958623076
MAE γ=0.01, C=0.01, ε=10: 6.260981365954933
MAE γ=0.01, C=0.01, ε=100: 8.403225806451614
MAE γ=0.01, C=0.01, ε=1000: 8.403225806451628
MAE γ=0.01, C=0.1, ε=0.001: 6.268344202633194
MAE γ=0.01, C=0.1, ε=0.01: 6.268344202633194
MAE γ=0.01, C=0.1, ε=0.1: 6.268564805552097
MAE γ=0.01, C=0.1, ε=1: 6.307797797332734
MAE γ=0.01, C=0.1, ε=10: 6.258200756323514
MAE γ=0.01, C=0.1, ε=100: 8.403225806451614
MAE γ=0.01, C=0.1, ε=1000: 8.403225806451628
MAE γ=0.01, C=1, ε=0.001: 6.205596020193872
MAE γ=0

In [105]:
# MAE γ=0.1, C=1000, ε=1: 3.219836232402966
for gamma in [.08, .09, 1, 1.1, 1.2]:
    for C in [800, 100, 1200, 1400]:
        for epsilon in [.8, .9, 1, 1.1, 1.2]:
            model = SVR(kernel='rbf', C=C, epsilon=epsilon, gamma=gamma)
            model.fit(X_train, y_train)

            y_pred = model.predict(X_val)

            print(f'MAE γ={gamma}, C={C}, ε={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE γ=0.08, C=800, ε=0.8: 3.252975373971557
MAE γ=0.08, C=800, ε=0.9: 3.2493138959133994
MAE γ=0.08, C=800, ε=1: 3.2506456926731087
MAE γ=0.08, C=800, ε=1.1: 3.241715269775642
MAE γ=0.08, C=800, ε=1.2: 3.2384944351914697
MAE γ=0.08, C=100, ε=0.8: 3.3793597615677005
MAE γ=0.08, C=100, ε=0.9: 3.387918118759331
MAE γ=0.08, C=100, ε=1: 3.4072864984627667
MAE γ=0.08, C=100, ε=1.1: 3.3899105192598835
MAE γ=0.08, C=100, ε=1.2: 3.3969268955754943
MAE γ=0.08, C=1200, ε=0.8: 3.287527546601245
MAE γ=0.08, C=1200, ε=0.9: 3.2657006243952775
MAE γ=0.08, C=1200, ε=1: 3.2482416184704737
MAE γ=0.08, C=1200, ε=1.1: 3.2469674188854523
MAE γ=0.08, C=1200, ε=1.2: 3.2486611371217116
MAE γ=0.08, C=1400, ε=0.8: 3.2803010546414284
MAE γ=0.08, C=1400, ε=0.9: 3.263029585209105
MAE γ=0.08, C=1400, ε=1: 3.2363943130259543
MAE γ=0.08, C=1400, ε=1.1: 3.22923850883603
MAE γ=0.08, C=1400, ε=1.2: 3.233270296816916
MAE γ=0.09, C=800, ε=0.8: 3.268839423084077
MAE γ=0.09, C=800, ε=0.9: 3.2519685635239166
MAE γ=0.09, C=800

In [106]:
# MAE γ=0.09, C=1400, ε=1.1: 3.1967806334584967
for gamma in [.09]:
    for C in [1400, 1500, 1600, 170]:
        for epsilon in [1.1, 1.2]:
            model = SVR(kernel='rbf', C=C, epsilon=epsilon, gamma=gamma)
            model.fit(X_train, y_train)

            y_pred = model.predict(X_val)

            print(f'MAE γ={gamma}, C={C}, ε={epsilon}: {mean_absolute_error(y_val, y_pred)}')

MAE γ=0.09, C=1400, ε=1.1: 3.1967806334584967
MAE γ=0.09, C=1400, ε=1.2: 3.2019576752587584
MAE γ=0.09, C=1500, ε=1.1: 3.1892656007546174
MAE γ=0.09, C=1500, ε=1.2: 3.1931039376136465
MAE γ=0.09, C=1600, ε=1.1: 3.18140250861775
MAE γ=0.09, C=1600, ε=1.2: 3.1854575778626946
MAE γ=0.09, C=170, ε=1.1: 3.258409928034243
MAE γ=0.09, C=170, ε=1.2: 3.284500130209703


Μετά από έλεγχο, παίρνω καλά αποτελέσματα για `γ=0.09, C=1600, ε=1.1` που αποδίδουν ΜΑΕ 3.1814. Δοκιμάζω το παραπάνω μοντέλο στα test data για να το συγκρίνω με τα υπόλοιπα:

In [107]:
best_rbf = SVR(kernel='rbf', gamma=.09, C=1600, epsilon=1.1)
print(f'Best SVR RBF MAE: {train_and_test(best_rbf)}')

Best SVR RBF MAE: 3.7546615665569716


Παρατηρούμε γενικά ότι το συγκεκριμένο dataset είναι σχετικά εύκολο. Δυνατές μηχανές, όπως RBF και πολυωνυμικές μεγάλου βαθμού δυσκολεύονται να ξεπεράσουν απλά μοντέλα όμως γραμμικό SVR και Linear Regressor.